src/test/java/com/artipie/api/ManageUsersTest.java

```diff
@@ -0,0 +1,182 @@
+import org.hamcrest.core.IsEqual;
+import org.json.JSONException;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.skyscreamer.jsonassert.JSONAssert;
+
+/**
+ * Test for {@link ManageUsers}.
+ * @since 0.1
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+class ManageUsersTest {
+
+    /**
+     * Test credentials key.
+     */
+    private static final Key KEY = new Key.From("creds.yaml");
+
+    /**
+     * Test storage.
+     */
+    private BlockingStorage blsto;
+
+    /**
+     * Test users.
+     */
+    private CrudUsers users;
+
+    @BeforeEach
+    void init() {
+        this.blsto = new BlockingStorage(new InMemoryStorage());
+        this.users = new ManageUsers(ManageUsersTest.KEY, this.blsto);
+    }
+
+    @Test
+    void listUsers() throws JSONException {
+        this.blsto.save(
+            ManageUsersTest.KEY,
+            new CredsConfigYaml().withUserAndGroups("Alice", List.of("readers"))
+                .withFullInfo(
+                    "Bob", Users.PasswordFormat.PLAIN, "xyz", "bob@example.com", Set.of("admin")
+                ).toString().getBytes(StandardCharsets.UTF_8)
+        );
+        JSONAssert.assertEquals(
+            this.users.list().toString(),
+            // @checkstyle LineLengthCheck (1 line)
+            "[{\"name\":\"Alice\",\"pass\":123,\"type\":\"plain\",\"groups\":[\"readers\"]},{\"name\":\"Bob\",\"type\":\"plain\",\"pass\":\"xyz\",\"email\":\"bob@example.com\",\"groups\":[\"admin\"]}]",
+            true
+        );
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void returnsEmptyList(final boolean add) {
+        if (add) {
+            this.blsto.save(ManageUsersTest.KEY, new byte[]{});
+        }
+        MatcherAssert.assertThat(
+            this.users.list(),
+            new IsEmptyCollection<>()
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("creds")
+    void addsUser(final Pair<YamlMapping, Boolean> pair) {
+        if (pair.getRight()) {
+            this.blsto.save(
+                ManageUsersTest.KEY,
+                pair.getLeft().toString().getBytes(StandardCharsets.UTF_8)
+            );
+        }
+        final String alice = "Alice";
+        final String email = "Alice@example.com";
+        final String pass = "xyz";
+        this.users.add(
+            Json.createObjectBuilder().add("type", "plain").add("pass", pass)
+                .add("email", email)
+                .add("groups", Json.createArrayBuilder().add("reader").add("creator").build())
+                .build(),
+            alice
+        );
+        final JsonArray list = this.users.list();
+        final JsonObject nuser = list.stream()
```

Comment:
Please add checking of availability of added user
Something like this:
        MatcherAssert.assertThat(
            "Failed to find added user",
            list.stream()
                .anyMatch(usr -> alice.equals(usr.asJsonObject().getString("name"))),
            new IsEqual<>(true)
        );